rewriter/RBSAssertions.cc

```diff
@@ -0,0 +1,509 @@
+
+        tree = ast::make_expression<ast::If>(iff->loc, move(iff->cond), move(iff->thenp), move(iff->elsep));
+    }
+
+    /*
+     * Splat expression such as:
+     *
+     *     x, y = value #: [Integer, String]
+     *
+     * are desugared using a Magic.expand-splat call:
+     *
+     *     <assignTemp>$1 = value
+     *     <assignTemp>$2 = ::<Magic>.<expand-splat>(<assignTemp>$1 ...)
+     *     x = <assignTemp>$2[0]
+     *     y = <assignTemp>$2[1]
+     *     <assignTemp>$1
+     *
+     * We need to transform this into:
+     *
+     *     <assignTemp>$1 = T.let(value, [Integer, String])
+     *     <assignTemp>$2 = ::<Magic>.<expand-splat>(<assignTemp>$1 ...)
+     *     x = <assignTemp>$2[0]
+     *     y = <assignTemp>$2[1]
+     *     <assignTemp>$1
+     */
+    void postTransformInsSeq(core::MutableContext ctx, ast::ExpressionPtr &tree) {
+        // Are we in an ins seq?
+        auto insSeq = ast::cast_tree<ast::InsSeq>(tree);
+        if (!insSeq) {
+            return;
+        }
+
+        // Do we have at least 2 stats?
+        // So we have the `<assignTemp>$1 = T.let...` and `<assignTemp>$2 = <Magic>.expand-splat(...)`
+        if (insSeq->stats.size() < 2) {
+            return;
+        }
+
+        // Is the first stat an assign to a desugared identifier?
+        auto assignTemp1 = ast::cast_tree<ast::Assign>(insSeq->stats[0]);
+        if (!assignTemp1 || !isDesugarIdent(ctx, assignTemp1->lhs)) {
+            return;
+        }
+
+        // Is the second stat also an assign to a desugared identifier?
+        auto assignTemp2 = ast::cast_tree<ast::Assign>(insSeq->stats[1]);
+        if (!assignTemp2 || !isDesugarIdent(ctx, assignTemp2->lhs)) {
+            return;
+        }
+
+        // Is the rhs of the second stat a call to `<Magic>.<expand-splat>`?
+        auto expandSplat = ast::cast_tree<ast::Send>(assignTemp2->rhs);
+        if (!expandSplat || expandSplat->fun != core::Names::expandSplat()) {
+            return;
+        }
+
+        // Is the receiver of the send the `<Magic>` constant?
+        auto recv = ast::cast_tree<ast::ConstantLit>(expandSplat->recv);
+        if (!recv || recv->symbol() != core::Symbols::Magic()) {
+            return;
+        }
+
+        // Transform the first assign into an `Assign` with a `Let` if it has an RBS comment
+        if (auto newAssign = transformAssign(ctx, assignTemp1)) {
+            insSeq->stats[0] = move(newAssign);
+        }
+
+        return;
+    }
+};
+
+/**
+ * Parse the type parameters from the previous statement
+ *
+ * Given a case like this one:
+ *
+ *     #: [X] (X) -> void
+ *     def foo(x)
+ *       y = nil #: X?
+ *     end
+ *
+ * We need to be aware of the type parameter `X` so we can use it to resolve the type of `y`.
+ */
+vector<pair<core::LocOffsets, core::NameRef>> parseTypeParams(core::MutableContext ctx, ast::ExpressionPtr *prevStat) {
```

Comment:
Should this be `const`?

```suggestion
vector<pair<core::LocOffsets, core::NameRef>> parseTypeParams(core::MutableContext ctx, const ast::ExpressionPtr *prevStat) {
```